_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+
+Notice lastly that extending Reservation does not bypass the constructor, so that the <strong>require(condition)</strong> will always be executed.  
+
+
+# 2. Hiding the public constructor of the class, defining a public one that validates and returns Option[Reservation]
+
+
+The idea here is to make it impossible for future users of Reservation to create an instance directly but only through one single method, engineered by us to take into account unlawful constructor parameters. The method ought to return a Reservation boxed in a data type that allows to represent something went wrong like the <code>Option</code>/<code>Try</code>/<code>Either</code>.  
+It is then the responsibility of the user to deal with the returned boxed type accordingly.
+
+There are two ways to achieve this. The sub chapter 19.2 in <a id="mmds-ref" href="#programming-in-scala">[1]</a> is useful in what follows.  
+
+
+The first is by adding <strong>private</strong> before the parameter list of the class, and then defining an <strong>apply</strong> method on the companion object.  
+
+```scala
+class Reservation[SeatType <: Seat] private (
+  userTicket: Ticket[SeatType],
+  guestTicket: Ticket[SeatType]
+)
+```
+
+While one does not need to name the method <strong>apply</strong>, it must be on the companion object; otherwise one does not have access to the constructor which is defined as private. Recall that the companion object has the same access rights as the class.
+
+```scala
+object Reservation {
+  def apply[SeatType <:Seat]( userTicket: Ticket[SeatType], guestTicket: Ticket[SeatType]): Option[Reservation[SeatType]] = (userTicket.seat, guestTicket.seat) match {
+    case (NormalSeat(_, _), NormalSeat(_, _)) => Some(new Reservation(userTicket, guestTicket))
+    case (VipTribune(_), VipTribune(_)) => Some(new Reservation(userTicket, guestTicket))
+    case (_: Standing.type , _: Standing.type ) => Some(new Reservation(userTicket, guestTicket))
+    case _ => None
+  }
+}
+```
+
+Notice that trying to extend Reservation on another file, possibly bypassing our <strong>apply</strong> method, will not work.
+
+As an alternative, one can hide the case class altogether, i.e. not only the constructor but also the name (so that we cannot use it as a type). To that end one defines a trait which exposes the implementation that we want.
+
+
+```scala
+sealed trait Reservation[SeatType <: Seat] {
+  val userTicket: Ticket[SeatType]
+  val guestTicket: Ticket[SeatType]
+}
+```
+
+and on the companion object we have the actual implementation:
```

Comment:
Please capitalize the "a" in "and" at the beginning of the sentence, since you're starting a new one.

Reply:
done